django/utils/html.py

```diff
@@ -153,11 +155,27 @@ def format_html_join(sep, format_string, args_generator):
       format_html_join('\n', "<li>{} {}</li>", ((u.first_name, u.last_name)
                                                   for u in users))
     """
-    return mark_safe(
-        conditional_escape(sep).join(
-            format_html(format_string, *args) for args in args_generator
+    if not isinstance(args_generator, Iterable):
+        raise TypeError("args_generator must be iterable.")
+
+    args_generator, _ = tee(args_generator)
+
+    # Check if args generator is an iterable of dict-likes or an iterable or iterables
+    if all(isinstance(arg, Mapping) for arg in _):
+        return mark_safe(
+            conditional_escape(sep).join(
+                format_html(format_string, **args) for args in args_generator
+            )
         )
```

Comment:
We don’t need `tee()` or a duplicated implementation of the outer loop. Instead, we can just do the check as we iterate, making the whole function body:

```python
return mark_safe(
    conditional_escape(sep).join(
        format_html(format_string, **args) if isinstance(args, Mapping) else format_html(format_string, *args)
        for args in args_generator
    )
)
```

Using `tee` and then checking the whole iterable with `all()` consumes the whole iterable “behind the scenes”, inside the `tee` object. Thus, it loses the any memory advantage when `args_generator` is a generator. You may as well have used a `list` cast, that would be more obvious what’s happening.

Reply:
Makes sense, I used `tee` and `all` to have a clearer Error message if the provided generator yields unsuitable items, but as you said it's not worth it because of the overhead.